docs/section-8-code-navigation-systems.md

```diff
@@ -0,0 +1,29 @@
+---
+title: Code Navigation Systems
+permalink: code-navigation-systems
+---
+
+# Code Navigation Systems
+
+Tree-sitter can be used in conjunction with its [tree query language](https://tree-sitter.github.io/tree-sitter/using-parsers#pattern-matching-with-queries) as a part of code navigation systems. An example of such a system can be seen in the `tree-sitter tag` command, which emits a textual dump of the interesting syntactic nodes in its file argument. This document exists to provide guidelines on the design and use of tree-sitter concepts to implement such systems.
+
+## Tagging and captures
+
+Code navigation systems using Tree-sitter need to use queries and captures to categorize and label different syntactic nodes, such as functions, built-ins, operators, and variables. A reverse-DNS-style notation is recommended for these captures, and provide guidelines below for naming captures of a given syntax node. User applications may extend (or only recognize a subset of) these capture names, but it is desirable to standardize on the names below when supported by a given system or language.
```

Comment:
I think we can expand on this a bit.  There are a few points we're trying to call out:

- "Tagging" is the act of identifying the entities that can be named in a program.  We use tree-sitter queries to find those entities.  Having found them, you use a syntax capture to label the entity and its name.

- There are two orthogonal things that we're tracking about each entity: what "kind" of thing it is (class, function, method, etc), and the "role" (better word?) for how the name is being used (just "definition" and "reference", could include "declaration" and maybe some others in the future).  Our convention is that you use a syntax capture that follows the pattern `@role.kind` for the entity itself, and another capture (always called `@name`) to pull out its name.

- By the way, here's a standardized vocabulary of the "roles" and "kinds", which can help language communities write tagging rules that will work out of the box with a wide variety of code analysis tools that need this information.